### ISSUE TITLE
TypeError: 'bool' object is not iterable when operating on schedule

### ISSUE THREAD

**Issue by OrangeDog** [saltstack/salt#51266]
When using the schedule module (e.g. applying `schedule.present` or running `schedule.enable`) a `TypeError` is thrown. The function returns successfully, but the schedule has not been persisted.
```
[ERROR   ] Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x7fee3181ebf8>, <tornado.concurrent.Future object at 0x7fee2972fc18>)
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/tornado/stack_context.py", line 277, in null_wrapper
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python3/dist-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python3/dist-packages/salt/minion.py", line 946, in handle_event
    yield [minion.handle_event(package) for minion in self.minions]
    future.result()
    ret = callback()
    return fn(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/tornado/ioloop.py", line 626, in _discard_future_result
  File "/usr/lib/python3/dist-packages/tornado/concurrent.py", line 238, in result
    value = future.result()
  File "<string>", line 4, in raise_exc_info
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1063, in run
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1055, in run
  File "/usr/lib/python3/dist-packages/salt/utils/schedule.py", line 381, in modify_job
    self.delete_job(name, persist)
    raise_exc_info(self._exc_info)
    result_list.append(f.result())
    raise_exc_info(self._exc_info)
  File "/usr/lib/python3/dist-packages/salt/minion.py", line 2135, in manage_schedule
    self.schedule.modify_job(name, schedule, persist)
  File "<string>", line 4, in raise_exc_info
  File "/usr/lib/python3/dist-packages/tornado/concurrent.py", line 238, in result
  File "<string>", line 4, in raise_exc_info
    yielded = next(result)
  File "/usr/lib/python3/dist-packages/salt/minion.py", line 2275, in handle_event
    self.manage_schedule(tag, data)
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 828, in callback
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 307, in wrapper
  File "/usr/lib/python3/dist-packages/salt/utils/schedule.py", line 257, in delete_job
    remove_hidden=True)
    for item in _schedule[job]:
    self.persist()
TypeError: 'bool' object is not iterable
  File "/usr/lib/python3/dist-packages/salt/utils/schedule.py", line 158, in _get_schedule
  File "/usr/lib/python3/dist-packages/salt/utils/schedule.py", line 222, in persist
```
Current contents of /etc/salt/minion.d/_schedule.conf:
```
schedule:
  __mine_interval: {enabled: true, function: mine.update, jid_include: true, maxrunning: 2,
    minutes: 60, return_job: false}
  highstate:
    enabled: true
    function: state.apply
    jid_include: true
    kwargs: {saltenv: base}
    maxrunning: 1
    minutes: 10
    name: highstate
    run_on_start: true
```
Versions report:
```
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.8
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: 0.26.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.26.2
         Python: 3.6.7 (default, Oct 22 2018, 11:32:17)
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5

System Versions:
           dist: Ubuntu 18.04 bionic
         locale: ISO-8859-1
        machine: x86_64
        release: 4.15.0-43-generic
         system: Linux
        version: Ubuntu 18.04 bionic
```

**Comment by garethgreenaway** [comment]
@OrangeDog Thanks for the report.  Unfortunately I haven't been able to duplicate this particular issue. Are you adding the schedule item via a state file?  If so, can you provide the contents of the state file you're using?  There was an issue in 2018.3 related to option of having the scheduler globally enabled or disabled, which was fixed here (https://github.com/saltstack/salt/pull/49019), but that particular issue should have been fixed for 2018.3.3.

**Comment by OrangeDog** [comment]
Sadly I can't reliably duplicate it either. I manually edited the `_schedule.conf` to resolve the minions that were stuck. It has also happened before though. It certainly does seem to involve a combination applying `schedule.present` states and en/disabling the global scheduler.

For this specific instance, I was adding a `returner` to this state.
```
highstate:
  schedule.present:
    - enabled: true
    - function: state.apply
    - job_kwargs:
        saltenv: base
    - minutes: 10
    - run_on_start: true
```

**Comment by tuxthepenguin84** [comment]
I'm having this issues as well. I don't recall what I did to get it in this state, but I am working with schedule.present and enabling/disabling the global scheduler.  At this point I'm not sure how to resolve it either, I've attempted manually updating the schedule.conf and reinstalling salt-minion as well.

Any suggestions?

**Comment by tuxthepenguin84** [comment]
More information:
Whatever the issue is, it's now happening on new salt-minion installs, which I do through the bootstrap installer.
I can duplicate the issue when I modify the global schedule disable/enable and then save it. When I save it is when the issues occurs.

Redacting some information for privacy.

Master Version Info
```
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 12 2018, 14:36:49)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-141-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

Minion Version Info:
```
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 12 2018, 14:36:49)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-141-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```
Output from salt-minion when it fails:
```
[INFO    ] Starting a new job with PID 15420
[DEBUG   ] LazyLoaded schedule.save
[DEBUG   ] LazyLoaded direct_call.execute
[DEBUG   ] SaltEvent PUB socket URI: /var/run/salt/minion/minion_event_8a2e12a902_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: /var/run/salt/minion/minion_event_8a2e12a902_pull.ipc
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/minion/minion_event_8a2e12a902_pub.ipc
[DEBUG   ] LazyLoaded event.fire
[DEBUG   ] SaltEvent PUB socket URI: /var/run/salt/minion/minion_event_8a2e12a902_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: /var/run/salt/minion/minion_event_8a2e12a902_pull.ipc
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/minion/minion_event_8a2e12a902_pull.ipc
[DEBUG   ] Sending event: tag = manage_schedule; data = {u'_stamp': '2019-02-01T17:31:38.490113', u'func': u'save_schedule'}
[DEBUG   ] Minion of 'REDACTED' is handling event tag 'manage_schedule'
[DEBUG   ] Persisting schedule
[ERROR   ] Exception in callback <functools.partial object at 0x7f9952f33050>
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/tornado/ioloop.py", line 592, in _run_callback
    ret = callback()
  File "/usr/lib/python2.7/dist-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/tornado/ioloop.py", line 598, in <lambda>
    self.add_future(ret, lambda f: f.result())
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 876, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 946, in handle_event
    yield [minion.handle_event(package) for minion in self.minions]
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 870, in run
    value = future.result()
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 656, in callback
    result_list.append(f.result())
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result)
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 2275, in handle_event
    self.manage_schedule(tag, data)
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 2155, in manage_schedule
    self.schedule.save_schedule()
  File "/usr/lib/python2.7/dist-packages/salt/utils/schedule.py", line 492, in save_schedule
    self.persist()
  File "/usr/lib/python2.7/dist-packages/salt/utils/schedule.py", line 222, in persist
    remove_hidden=True)
  File "/usr/lib/python2.7/dist-packages/salt/utils/schedule.py", line 158, in _get_schedule
    for item in _schedule[job]:
TypeError: 'bool' object is not iterable
[INFO    ] User REDACTED Executing command saltutil.find_job with jid 20190201093143568582
[DEBUG   ] Command details {u'tgt_type': u'list', u'jid': u'20190201093143568582', u'tgt': [u'REDACTED'], u'ret': u'',                                 u'user': u'REDACTED', u'arg': [u'20190201093138459961'], u'fun': u'saltutil.find_job'}
[INFO    ] Starting a new job with PID 15423
[DEBUG   ] LazyLoaded saltutil.find_job
[DEBUG   ] LazyLoaded direct_call.execute
[DEBUG   ] Minion return retry timer set to 10 seconds (randomized)
[INFO    ] Returning information for job: 20190201093143568582
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'/etc/salt/pki/minion', u'REDACTED', u'tcp://REDACTED:450                                6', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'/etc/salt/pki/minion', u'REDACTED', u'tcp://REDACTED:4506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://REDACTED:4506
[DEBUG   ] Trying to connect to: tcp://REDACTED:4506
[DEBUG   ] minion return: {u'fun_args': [u'20190201093138459961'], u'jid': u'20190201093143568582', u'return': {'tgt_type': 'glob', 'jid':                                 '20190201093138459961', 'tgt': 'REDACTED*', 'pid': 15420, 'ret': '', 'user': 'REDACTED', 'arg': [], 'fun': 'schedule.save'},                                 u'retcode': 0, u'success': True, u'fun': u'saltutil.find_job'}
^C[WARNING ] Minion received a SIGINT. Exiting.
[INFO    ] Shutting down the Salt Minion
The Salt Minion is shutdown. Minion received a SIGINT. Exited.
[DEBUG   ] Minion return retry timer set to 6 seconds (randomized)
[INFO    ] Returning information for job: 20190201093138459961
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'/etc/salt/pki/minion', u'REDACTED', u'tcp://REDACTED:4506', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'/etc/salt/pki/minion', u'REDACTED', u'tcp://REDACTED:4506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://REDACTED:4506
[DEBUG   ] Trying to connect to: tcp://REDACTED:4506
[DEBUG   ] minion return: {u'fun_args': [], u'jid': u'20190201093138459961', u'return': {u'comment': u'Failed to save schedule.', u'result': False}, u'retcode': 0, u'success': True, u'fun': u'schedule.save'}
```
Output of the salt-master:
```
root@REDACTED:/#
root@REDACTED:/# salt 'REDACTED*' schedule.list
REDACTED.REDACTED:
    ----------
    schedule:
        ----------
root@REDACTED:/# salt 'REDACTED*' schedule.enable
REDACTED.REDACTED:
    ----------
    comment:
        Enabled schedule on minion.
    result:
        True
root@REDACTED:/# salt 'REDACTED*' schedule.list
REDACTED.REDACTED:
    schedule:
      enabled: true
root@REDACTED:/# salt 'REDACTED*' schedule.save
REDACTED.REDACTED:
    ----------
    comment:
        Failed to save schedule.
    result:
        False
root@REDACTED:/#
```

**Comment by garethgreenaway** [comment]
@tuxthepenguin84 Thanks for that note, running `schedule.save` was the missing piece of the puzzle.  I've identified the issue and have a PR ready to go that fixes the issue.  Also reported in, https://github.com/saltstack/salt/issues/51600

**Comment by garethgreenaway** [comment]
Closing this out as the fix is in the 2018.3 and later branches, and will be available in the next minor releases.  If the problem persists, let us know and we can reopen this issue.
